### PR TITLE
Addition of the hydrostatic component

### PR DESCRIPTION
**Pull request by rubenzorrilla**
Addition of the hydrostatic pressure component to the new outlet process. If selected, the hydrostatic pressure component coming from the body force is added to both the PRESSURE and the EXTERNAL_PRESSURE variables.

It is also possible to consider the rectangular extra pressure coming from a fluid mass over the outlet setting the "h_top" field (@jginternational is working in the interface to hide this value if hydrostatic outlet is not considered).

### CODE REVIEW

**Comment by jginternational** [comment]
Interface branch related to this issue is ready to merge

https://github.com/KratosMultiphysics/GiDInterface/issues/61

**Comment by rubenzorrilla** [comment]
A check has been added to avoid division by zero when the hydrostatic outlet is set with zero value body force. Should we throw an error in this case¿ (the hydrostatic component will always be zero) @maceligueta @RiccardoRossi

File: applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py [hunk 1]
```diff
@@ -68,26 +80,63 @@ def ExecuteInitializeSolutionStep(self):
...
+            node.SetSolutionStepValue(KratosMultiphysics.EXTERNAL_PRESSURE, 0, hyd_pres+cur_pres)               # Add the hydrostatic component to the EXTERNAL_PRESSURE value
+
+
+    def _ComputeOutletCharacteristicVelocity(self):
+        # Compute the current process average velocity
+        outlet_avg_vel_norm = 0.0
+        for node in self.outlet_model_part.GetCommunicator().LocalMesh().Nodes:
+            vnode = node.GetSolutionStepValue(KratosMultiphysics.VELOCITY)
+            outlet_avg_vel_norm += math.sqrt(vnode[0]**2+vnode[1]**2+vnode[2]**2)
+            outlet_avg_vel_norm /= len(self.outlet_model_part.GetCommunicator().LocalMesh().Nodes)
```
Comment (maceligueta) [inline applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py]: `math.sqrt(vnode[0]*vnode[0]+vnode[1]*vnode[1]+vnode[2]*vnode[2])`  maybe would be faster here

File: applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py [hunk 2]
```diff
@@ -68,26 +80,63 @@ def ExecuteInitializeSolutionStep(self):
...
+
+    # Private methods section
+    def _AddOutletHydrostaticComponent(self):
+        # Get the body force value
+        for node in self.outlet_model_part.Nodes:
+            body_force = node.GetSolutionStepValue(KratosMultiphysics.BODY_FORCE, 0)
+            break
+
+        # Compute the body force unit normal vector
+        body_force_norm = math.sqrt(math.pow(body_force[0],2)+math.pow(body_force[1],2)+math.pow(body_force[2],2))  # Body force norm
```
Comment (maceligueta) [inline applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py]: `math.pow(body_force[0],2)` can be replaced by `body_force[0]*body_force[0]`  (maybe it wil be faster)

**Review by maceligueta — APPROVED** [review]
Personally, I prefer to implement all loops over Nodes in c++. I noticed a big difference in time consumption. But still the code looks fine to me.

**Comment by RiccardoRossi** [comment]
Cantidad this be bases on the proceses un the Core with local axes?

That way It Will be done in c++ without extra code

El 7 abr. 2017 11:45 a. m., "Miguel Angel" <notifications@github.com>
escribió:

*@maceligueta* approved this pull request.

Personally, I prefer to implement all loops over Nodes in c++. I noticed a
big difference in time consumption. But still the code looks fine to me.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/KratosMultiphysics/Kratos/pull/259#pullrequestreview-31521528>,
or mute the thread

**Comment by rubenzorrilla** [comment]
@maceligueta I noticed such a big difference too... If you all agree we can create a c++ outlet process together with its Python interface.

@RiccardoRossi Sorry but I did not understand your point ...

File: applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py (hunk 1 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py]: After a short research, I found out that for low values of "n" naive product x*x is cheaper than x**n or pow(x,n). Let me change it before merging.

File: applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py (hunk 2 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/python_scripts/apply_outlet_process.py]: See above.

**Comment by maceligueta** [comment]
I think @RiccardoRossi is saying that some operations of this process can be done by calling a process in the core. And the process in the core is already implemented in c++...

**Comment by rubenzorrilla** [comment]
After having a look, I did not find out any core process that can be used here... If you do not disagree, I will merge the changes since the interface has been already modified. We can open a new branch in order to optimize these operations.

**Pull request merged**